File: trunk/CodeModel/Language/AS3/AS3Package.cs
/*
Copyright (c) 2008 Joa Ebert

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

using System;
using System.Collections.Generic;
using System.Text;
using AS3V.CodeModel.Generic;
using System.IO;
using AS3V.CodeModel.Exceptions;
using AS3V.CodeModel.Generic.Enumerators;

namespace AS3V.CodeModel.Language.AS3
{
    public class AS3Package : IPackage
    {
        private string _name;
        private string _path;

        private AS3Project _project;
        private AS3Package _parent;
        
        private List<AS3Package> _packages;
        private List<AS3CompilationUnit> _units;

        public AS3Package()
        {
            _name = "";
            _path = "";

            _project = null;
            _parent = null;

            _packages = new List<AS3Package>();
            _units = new List<AS3CompilationUnit>();
        }

        public void Build(AS3Project project, AS3Package parent, string path)
        {
            _project = project;
            _parent = parent;

            _path = path;
            _name = new FileInfo(path).Name;

            string[] dirs = Directory.GetDirectories(path);
            string[] files = Directory.GetFiles(path);

            int i, n;

            for (i = 0, n = dirs.Length; i < n; ++i)
            {
                AS3Package package = new AS3Package();

                package.Build(project, this, dirs[i]);

                _packages.Add(package);
            }

            for (i = 0, n = files.Length; i < n; ++i)
            {
                AS3CompilationUnit unit = new AS3CompilationUnit();

                unit.Build(project, this, files[i]);

                _units.Add(unit);
            }
        }

        #region IPackage Member

        public virtual string Name
        {
            get { return _name; }
        }

        public string DirectoryPath
        {
            get { return _path; }
        }

        public virtual IPackage Parent
        {
            get { return _parent;  }
        }

        public IProject Project
        {
            get { return _project; }
        }

        public int PackageCount
        {
            get { return _packages.Count; }
        }

        public int CompilationUnitCount
        {
            get { return _units.Count; }
        }

        public IPackage PackageAt(int index)
        {
            if (index < 0 || index >= PackageCount)
            {
                throw new RangeException(index, PackageCount);
            }

            return _packages[index];
        }

        public ICompilationUnit CompilationUnitAt(int index)
        {
            if (index < 0 || index >= CompilationUnitCount)
            {
                throw new RangeException(index, CompilationUnitCount);
            }

            return _units[index];
        }

        public virtual string PackagePath
        {
            get
            {
                string result = Name;

                IPackage package = Parent;

                while (package != null)
                {
                    if (package.Name.Length > 0)
                    {
                        result = package.Name + "." + result;
                    }
                    else
                    {
                        break;
                    }

                    package = package.Parent;
                }

                return result;
            }
        }

        public IEnumerable<IPackage> PackageEnumerator
        {
            get { return new PackageEnumerator(this); }
        }

        public IEnumerable<ICompilationUnit> CompilationUnitEnumerator
        {
            get { return new CompilationUnitEnumerator(this); }
        }

        public bool HasPackage(string name)
        {
            return null != Package(name);
        }

        public IPackage Package(string name)
        {
            foreach (AS3Package package in PackageEnumerator)
            {
                if (package.Name == name)
                {
                    return package;
                }
            }

            return null;
        }

        public bool HasCompilationUnit(string name)
        {
            return null != CompilationUnit(name);
        }

        public ICompilationUnit CompilationUnit(string name)
        {
            foreach (AS3CompilationUnit unit in CompilationUnitEnumerator)
            {
                if (unit.Name == name)
                {
                    return unit;
                }
            }

            return null;
        }

        #endregion
    }
}

File: java/src/com/joa_ebert/as3v/language/as3/AS3Package.java
package com.joa_ebert.as3v.language.as3;

import java.io.File;
import java.util.ArrayList;
import java.util.Iterator;

import com.joa_ebert.as3v.projectModel.ICompilationUnit;
import com.joa_ebert.as3v.projectModel.IPackage;
import com.joa_ebert.as3v.projectModel.IProject;

public class AS3Package implements IPackage {

	private String _name = "";
	private String _path = "";
	private String _packagePath = null;
	private AS3Project _project = null;

	private AS3Package _parent = null;
	private ArrayList<IPackage> _packages = new ArrayList<IPackage>();
	private ArrayList<ICompilationUnit> _units = new ArrayList<ICompilationUnit>();

	public AS3Package() {
	}

	protected void build(AS3Project project, AS3Package parent,
			String fileSystemPath) {
		File packageFile = new File(fileSystemPath);

		_project = project;
		_parent = parent;
		_path = fileSystemPath;
		_name = packageFile.getName();

		String files[] = packageFile.list();

		int i = 0;
		int n = files.length;

		for (; i < n; ++i) {
			File child = new File(packageFile.getAbsolutePath()
					+ File.separator + files[i]);

			if (child.isDirectory()) {
				AS3Package package$ = new AS3Package();

				package$.build(project, this, child.getAbsolutePath());

				_packages.add(package$);
			} else {
				if (child.getAbsolutePath().endsWith(".as")) {
					AS3CompilationUnit unit = new AS3CompilationUnit();

					unit.build(project, this, child.getAbsolutePath());

					_units.add(unit);
				}
			}
		}
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public ICompilationUnit compilationUnit(String name) {
		for (Iterator<ICompilationUnit> iter = compilationUnitIterator(); iter
				.hasNext();) {
			ICompilationUnit unit = iter.next();

			if (unit.name().equals(name))
				return unit;
		}

		return null;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public ICompilationUnit compilationUnitAt(int index) {
		if (index < 0 || index >= _units.size())
			return null;
		return _units.get(index);
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public int compilationUnitCount() {
		return _units.size();
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Iterator<ICompilationUnit> compilationUnitIterator() {
		return _units.iterator();
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public String fileSystemPath() {
		return _path;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public boolean hasCompilationUnit(String name) {
		return null != compilationUnit(name);
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public boolean hasPackage(String name) {
		return null != package$(name);
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public String name() {
		return _name;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public IPackage package$(String name) {
		for (Iterator<IPackage> iter = packageIterator(); iter.hasNext();) {
			IPackage package$ = iter.next();

			if (package$.name().equals(name))
				return package$;
		}

		return null;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public IPackage packageAt(int index) {
		if (index < 0 || index >= _packages.size())
			return null;
		return _packages.get(index);
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public int packageCount() {
		return _packages.size();
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Iterator<IPackage> packageIterator() {
		return _packages.iterator();
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public String packagePath() {
		if (null != _packagePath)
			return _packagePath;

		String result = name();

		IPackage package$ = parent();

		while (null != package$) {
			if (package$.name().length() > 0) {
				result = package$.name() + "." + result;
			} else {
				break;
			}

			package$ = package$.parent();
		}

		_packagePath = result;
		
		return result;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public IPackage parent() {
		return _parent;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public IProject project() {
		return _project;
	}

}
